### PR TITLE
First token lm_head optimization

### PR DESCRIPTION
**Pull request by cyita**
## Description

Optimize the first token lm_head latency & memory usage.

### 1. Why the change?

Optimize the first token lm_head latency & memory usage.

### 2. User API changes

export BIGDL_OPTIMIZE_LM_HEAD=1 to use this optimization for llama & gptj.

### 4. How to test?
- [x] local test cpu
- [x] local test gpu

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/low_bit_linear.py [hunk 1]
```diff
@@ -535,7 +537,15 @@ def forward(self, x: torch.Tensor):
...
             self.bias.data = self.bias.data.to(x.dtype)
-
+        
+        if self.lm_head:
+            if x.dim() > 3:
+                x = x.reshape([-1, x.shape[-2], x.shape[-1]])
+            shape = list(x.size())
+            if shape[1] > 10:
+                shape[1] = 1
+                x = x[:, -1, :].view(shape)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/low_bit_linear.py:547]: Put in a utility function

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 2]
```diff
@@ -211,6 +211,11 @@ def _replace_with_low_bit_linear(model, qtype, modules_to_not_convert=None,
             if (not any(key in ".".join(current_key_name) for key in modules_to_not_convert) and
                     not isinstance(module, LowBitLinear)):
                 in_features, out_features, mp_group = linear_args
+                lm_head = False
+                if name == "lm_head":
+                    from bigdl.llm.transformers.utils import get_optimize_lm_head
+                    if model_type in ["gptj", "llama"] and get_optimize_lm_head():
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:217]: only applicable to "gptj" and "llama"?

Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/convert.py:217]: Not all model inputs dim >=3, and seq_len is not always the first dimension. Other models need to be validated before being included.

**Comment by jason-dai** [comment]
@hkvision - please test its impact on MTL

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 3]
```diff
@@ -211,6 +211,11 @@ def _replace_with_low_bit_linear(model, qtype, modules_to_not_convert=None,
             if (not any(key in ".".join(current_key_name) for key in modules_to_not_convert) and
                     not isinstance(module, LowBitLinear)):
                 in_features, out_features, mp_group = linear_args
+                optimize_lm_head = False
+                if name == "lm_head":
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/convert.py:215]: Not all models have name lm_head? seems chatglm has name output_layer?

File: python/llm/src/bigdl/llm/transformers/utils.py [hunk 4]
```diff
@@ -168,6 +168,19 @@ def get_ipex_version():
     return _ipex_version
 
 
+_optimize_lm_head = None
+
+
+def get_optimize_lm_head():
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/utils.py:174]: no need this? just `and os.environ.get("BIGDL_OPTIMIZE_LM_HEAD", None) == 1`?

**Review by hkvision — APPROVED** [review]

**Comment by hkvision** [comment]
> @hkvision - please test its impact on MTL

Sure.

File: python/llm/src/bigdl/llm/transformers/convert.py (hunk 3 above)
Comment (cyita) [reply python/llm/src/bigdl/llm/transformers/convert.py:215]: Yes, we need to check both seq_len dim and the name of lm_head before including new models.

**Comment by hkvision** [comment]
> @hkvision - please test its impact on MTL

Results here: https://github.com/analytics-zoo/nano/issues/1080#issuecomment-1980163619

**Comment by cyita** [comment]
The test results indicate that there is little performance improvement from this optimization on MTL.

**Pull request merged**